rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala

```diff
@@ -68,59 +68,85 @@ object SystemProcesses {
 
       type Channels = Seq[Result[ListParWithRandomAndPhlos]]
 
+      type Producer = (Seq[Par], Par) => F[Unit]
+
       private val prettyPrinter = PrettyPrinter()
 
       private val UNLIMITED_MATCH_PHLO = matchListPar(Cost(Integer.MAX_VALUE))
 
-      private def foldResult(
-          produceResult: Either[InterpreterError, Option[(ContWithMetaData, Channels)]]
-      ): F[Unit] =
-        produceResult.fold(
-          _ => F.raiseError(OutOfPhlogistonsError),
-          _.fold(F.unit) {
-            case (cont, channels) =>
-              dispatcher.dispatch(unpackCont(cont), channels.map(_.value), cont.sequenceNumber)
+      object ContractCall {
+        private def foldResult(
+            produceResult: Either[InterpreterError, Option[(ContWithMetaData, Channels)]]
+        ): F[Unit] =
+          produceResult.fold(
+            _ => F.raiseError(OutOfPhlogistonsError),
+            _.fold(F.unit) {
+              case (cont, channels) =>
+                dispatcher.dispatch(unpackCont(cont), channels.map(_.value), cont.sequenceNumber)
+            }
+          )
+
+        def unapply(contractArgs: (Seq[ListParWithRandomAndPhlos], Int)): Option[
+          (Producer, Seq[Par])
+        ] =
+          contractArgs match {
+            case (
+                Seq(
+                  ListParWithRandomAndPhlos(
+                    args,
+                    rand,
+                    _
+                  )
+                ),
+                sequenceNumber
+                ) =>
+              def produce(values: Seq[Par], ch: Par): F[Unit] =
+                for {
+                  produceResult <- space.produce(
+                                    ch,
+                                    ListParWithRandom(values, rand),
+                                    persist = false,
+                                    sequenceNumber
+                                  )(UNLIMITED_MATCH_PHLO)
+                  _ <- foldResult(produceResult)
+                } yield ()
+
+              Some((produce, args))
```

Comment:
It's very strange for an extractor to always return `Some`. I guess we should be returning `case _ => None`

Reply:
fixed